feat: ghc 9.12.2, cabal 3.14.1.1 and stack 3.1.1 on bookworm

jhrcek opened the pull request:
There's a question of what do do with all those buster images (since buster is long past the end of life).
For now I commented them out in CI config.
But I think we could even remove all of them from the repo. WDYT @develop7 ?

This PR aims to supersede https://github.com/haskell/docker-haskell/pull/145

develop7 reviewed (changes requested):
While 9.12 is more than welcome, phasing buster out of CI doesn't belong this PR IMO. Could you move it elsewhere first?

jhrcek commented:
@develop7 you're right. I removed all unrelated CI changes from this PR.
I removed them because I was just frustrated because I didn't understand where the failing jobs are coming from.
It was because I was missing the exclusions for 9.12 which doesn't have buster/bullseye variants.

But in general are you in favor of both removing buster from CI as well as removing all buster dockerfiles from the repo?
My arguments for removal: 1. don't waste CI resources building something that's past the EOL. 2. make the repo easier to navigate 3. if anyone wants those buster images they can find them in git history.

develop7 reviewed (approved):
LGTM, ship it
